src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,116 @@
+     */
+    static std::vector<std::function<std::unique_ptr<TypeToCreate>()>> getRegisteredConstructors();
+
+protected:
+    /**
+     * Adds a generic creator function to the registry
+     *
+     * @param generic_name The name of the generic creator function to be added
+     * @param generic_creator A "create" function that takes no arguments and will return
+     * a unique_ptr to a new instance of the specified generic type
+     */
+    static void registerCreator(
+            std::string generic_name,
+            std::function<std::unique_ptr<TypeToCreate>()> generic_creator);
+
+private:
+    /**
+     * Returns a reference to the generic registry. The registry is a map of Generic names
+     * to a "create" function that will create and return a unique_ptr to a new concrete
+     * instance of the generic type, which allows the code to be aware
+     * of all the Generics that are available.
+     *
+     * This is the same as the above public getRegistry function. We need a mutable
+     * version in order to add entries to the registry. The function is private so that
+     * only this class can make the modifications. Outside sources should not have direct
+     * access to modify the registry.
+     *
+     * @return a mutable reference to the generic registry
+     */
+    static GenericRegistry<IndexType, TypeToCreate>& getMutableRegistry();
+};
+
+/**
+ * This templated generic factory class is used by generic types that are derived from the
+ * Abstract TypeToCreate template class. Its purpose is to create a Factory for the implemented
+ * generic type and automatically register the Generic type in the GenericFactory registry.
+ *
+ * Declaring the static variable will also cause it to be initialized at the start of the
+ * program (because it's static). This will immediately call the constructor, which adds
+ * the backend T to the GenericFactory registry. From then on, the rest of the program
+ * can use the registry to find all the Backends that are available (and register with
+ * this templated class).
+ *
+ * @tparam T The class of the generic type neric to be added to the registry. For example, to add a
+ * new class called MoveBackend that inherits from Backend (or any other Generic), the
+ * following line should be added to the end of the .cpp file (without the quotations):
+ * "static TGenericFactory<MoveBackend> factory;"
+ */
+template <class IndexType, class TypeToCreate, class T>
+class TGenericFactory : public GenericFactory<IndexType, TypeToCreate>
+{
+    // compile time type checking that T is derived class of Generic
+    static_assert(std::is_base_of<TypeToCreate, T>::value,
+                  "T must be derived class of TypeToCreate!");
+
+public:
+    TGenericFactory()
+    {
+        auto generic_creator = []() -> std::unique_ptr<TypeToCreate> {
+            return std::make_unique<T>();
+        };
+        GenericFactory<IndexType, TypeToCreate>::registerCreator(T::name, generic_creator);
```

Comment:
@jonathanlew @akhilveeraghanta @MathewMacDougall - so right now this depends on the class having a public `name` member. We can definitely merge it like this, it will work as a replacement for all the factories in the code, but we might want to consider changing it to be more generic. Perhaps require that `T` implement a function `IndexType getFactoryIndex()`?

Reply:
I like that idea, I'd like to get away from static variables with names if possible. @garethellis0 can you make a ticket for this?

Reply:
Yeah no problem. Created #1142. @kelvinkoon - can you leave a comment `TODO (Issue #1142): Change to use a function instead of a static variable` here?

Reply:
Comment added.